File: web/src/types/account.ts
import { z } from 'zod'
import { Reference, referenceSchema } from '@/types/common/reference'
import { Amount } from '@/types/common/amount'

export enum AccountType {
  ACCOUNT = 'ACCOUNT',
  EXPENSE = 'EXPENSE',
  INCOME = 'INCOME',
}

export interface AccountReference extends Reference {
  type: AccountType
  reviseDate?: string
}

export const accountReferenceSchema = referenceSchema.extend({
  type: z.enum(AccountType),
  reviseDate: z.string().optional(),
})

export interface AccountGroup {
  id?: string
  name: string
  deleted: boolean
}

export interface Account {
  id?: string
  name: string
  type: AccountType
  parser?: string
  group?: Reference
  deleted: boolean
  reviseDate?: string
  reportExclude: boolean
  externalId?: string
}

export interface Currency {
  id?: string
  name: string
  crypto: boolean
}

export interface AccountBalance {
  account: AccountReference
  balances: Amount[]
}

export interface AccountBalanceGroup {
  id?: string
  name?: string
  accounts: AccountBalance[]
}

File: server/src/main/kotlin/com/evgenltd/financemanager/account/record/account.kt
package com.evgenltd.financemanager.account.record

import com.evgenltd.financemanager.account.entity.AccountType
import com.evgenltd.financemanager.common.util.Amount
import java.time.LocalDate
import java.util.*

data class AccountRecord(
    val id: UUID?,
    val name: String,
    val type: AccountType,
    val parser: String?,
    val deleted: Boolean,
    val reviseDate: LocalDate?,
    val reportExclude: Boolean = false,
    val externalId: String? = null,
)

data class AccountReferenceRecord(
    val id: UUID,
    val name: String,
    val deleted: Boolean,
    val type: AccountType,
    val reviseDate: LocalDate?,
)

data class AccountBalanceFilter(
    val hideZeroBalances: Boolean = false,
)

data class AccountBalanceRecord(
    val account: AccountReferenceRecord,
    val balances: List<Amount>,
)
